File: scripts/models/player_settings.lua
local constants = require "scripts.constants"

---@class C2CC.PlayerSettingsData
---@field NegativeSignals boolean Automatically make output item and fluid signals negative.
---@field Priority integer Default station priority value.
---@field DefaultNetworkSignal SignalID Default network mask signal prototype.
---@field NetworkFlag integer Default network mask bitmask value.
---@field Stacks integer Default output stacks count value.
---@field Count integer Default output items count value.
---@field DefaultInputMode string Default focused input mode ("count" or "stacks").

---@class C2CC.PlayerSettings : C2CC.PlayerSettingsData
local PlayerSettings = {}
PlayerSettings.__index = PlayerSettings

---@param playerIndex integer
---@return C2CC.PlayerSettings
function PlayerSettings.Get(playerIndex)
  storage.player_settings = storage.player_settings or {}
  local st = storage.player_settings[playerIndex]

  if not st then
    st = {}
    storage.player_settings[playerIndex] = st
  end

  if st.NegativeSignals == nil then st.NegativeSignals = true end
  if st.Priority == nil then st.Priority = constants.SETTINGS.DEFAULT_PRIORITY end
  if st.DefaultNetworkSignal == nil then st.DefaultNetworkSignal = constants.SETTINGS.DEFAULT_NETWORK_SIGNAL end
  if st.NetworkFlag == nil then st.NetworkFlag = constants.SETTINGS.DEFAULT_NETWORK_FLAG end
  if st.Stacks == nil then st.Stacks = constants.SETTINGS.DEFAULT_STACKS end
  if st.Count == nil then st.Count = constants.SETTINGS.DEFAULT_COUNT end
  if st.DefaultInputMode == nil then st.DefaultInputMode = constants.SETTINGS.DEFAULT_INPUT_MODE end

  setmetatable(st, PlayerSettings)
  return st
end

---@param playerIndex integer
---@return C2CC.PlayerSettings
function PlayerSettings.Copy(playerIndex)
  local src = PlayerSettings.Get(playerIndex)
  local draft = {}
  for k, v in pairs(src) do
    draft[k] = v
  end
  setmetatable(draft, PlayerSettings)
  return draft
end

---@param data table
function PlayerSettings:Update(data)
  if not data then return end
  if data.NegativeSignals ~= nil then self.NegativeSignals = data.NegativeSignals ~= false end
  if data.Priority ~= nil then self.Priority = tonumber(data.Priority) or constants.SETTINGS.DEFAULT_PRIORITY end
  if data.DefaultNetworkSignal ~= nil and data.DefaultNetworkSignal.name then self.DefaultNetworkSignal = data.DefaultNetworkSignal end
  if data.NetworkFlag ~= nil then self.NetworkFlag = tonumber(data.NetworkFlag) or constants.SETTINGS.DEFAULT_NETWORK_FLAG end
  if data.Stacks ~= nil then self.Stacks = tonumber(data.Stacks) or 0 end
  if data.Count ~= nil then self.Count = tonumber(data.Count) or 0 end
  if data.DefaultInputMode ~= nil then self.DefaultInputMode = tostring(data.DefaultInputMode) end
end

---@param playerIndex integer
---@param settingsData table
function PlayerSettings.Set(playerIndex, settingsData)
  if not playerIndex or not settingsData then return end
  local ps = PlayerSettings.Get(playerIndex)
  ps:Update(settingsData)
end

return PlayerSettings

File: src/scripts/models/player_settings.ts
import { SETTINGS } from '../constants';
import type { SignalID, PlayerIndex } from 'factorio:runtime';

export class PlayerSettings implements PlayerSettingsData {
  public negativeSignals: boolean = SETTINGS.DEFAULT_NEGATIVE_SIGNALS;
  public priority: Priority = SETTINGS.DEFAULT_PRIORITY;
  public defaultNetworkSignal?: SignalID = SETTINGS.DEFAULT_NETWORK_SIGNAL;
  public networkFlag: NetworkMask = SETTINGS.DEFAULT_NETWORK_FLAG;
  public stacks: number = SETTINGS.DEFAULT_STACKS;
  public count: number = SETTINGS.DEFAULT_COUNT;
  public defaultInputMode: string = SETTINGS.DEFAULT_INPUT_MODE;
  public autoQueryPriorities: boolean = true;

  constructor(init?: Partial<PlayerSettingsData>) {
    if (init) {
      Object.assign(this, init);
    }
  }

  public static get(playerIndex: PlayerIndex): PlayerSettings {
    const raw = storage.player_settings?.[playerIndex];
    const res = new PlayerSettings();
    if (raw) {
      res.negativeSignals = raw.negativeSignals ?? SETTINGS.DEFAULT_NEGATIVE_SIGNALS;
      res.priority = raw.priority ?? SETTINGS.DEFAULT_PRIORITY;
      res.defaultNetworkSignal = raw.defaultNetworkSignal ?? SETTINGS.DEFAULT_NETWORK_SIGNAL;
      res.networkFlag = raw.networkFlag ?? SETTINGS.DEFAULT_NETWORK_FLAG;
      res.stacks = raw.stacks ?? SETTINGS.DEFAULT_STACKS;
      res.count = raw.count ?? SETTINGS.DEFAULT_COUNT;
      res.defaultInputMode = raw.defaultInputMode ?? SETTINGS.DEFAULT_INPUT_MODE;
      res.autoQueryPriorities = raw.autoQueryPriorities ?? true;
    }
    return res;
  }

  public static copy(playerIndex: PlayerIndex): PlayerSettings {
    return new PlayerSettings(PlayerSettings.get(playerIndex));
  }

  public update(data?: Partial<PlayerSettingsData>): void {
    if (!data) return;
    if (data.negativeSignals !== undefined) this.negativeSignals = data.negativeSignals;
    if (data.priority !== undefined)
      this.priority = tonumber(data.priority) || SETTINGS.DEFAULT_PRIORITY;
    if (data.defaultNetworkSignal?.name) this.defaultNetworkSignal = data.defaultNetworkSignal;
    if (data.networkFlag !== undefined)
      this.networkFlag = tonumber(data.networkFlag) || SETTINGS.DEFAULT_NETWORK_FLAG;
    if (data.stacks !== undefined) this.stacks = tonumber(data.stacks) || 0;
    if (data.count !== undefined) this.count = tonumber(data.count) || 0;
    if (data.defaultInputMode !== undefined)
      this.defaultInputMode = tostring(data.defaultInputMode);
    if (data.autoQueryPriorities !== undefined) this.autoQueryPriorities = data.autoQueryPriorities;
  }

  public static set(playerIndex: PlayerIndex, settingsData: Partial<PlayerSettingsData>): void {
    if (!playerIndex || !settingsData) return;
    if (!storage.player_settings) storage.player_settings = {};
    const current = PlayerSettings.get(playerIndex);
    current.update(settingsData);
    storage.player_settings[playerIndex] = {
      negativeSignals: current.negativeSignals,
      priority: current.priority,
      defaultNetworkSignal: current.defaultNetworkSignal,
      networkFlag: current.networkFlag,
      stacks: current.stacks,
      count: current.count,
      defaultInputMode: current.defaultInputMode,
      autoQueryPriorities: current.autoQueryPriorities,
    };
  }
}
